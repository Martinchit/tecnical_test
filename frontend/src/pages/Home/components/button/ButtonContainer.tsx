import styled from 'styled-components';
import { Button } from 'react-bootstrap';

export const ButtonContainer = styled(Button)`
  background-color: #fff;
  border: none;
  &:hover {
    outline: none;
    box-shadow: none;
    background-color: #fff;
  }
  &:focus {
    outline: none;
    box-shadow: none;
    background-color: #fff;
  }
`;
