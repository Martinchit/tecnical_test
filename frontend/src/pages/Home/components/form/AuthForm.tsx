import React from 'react';
import { Button, Form } from 'react-bootstrap';
import { AuthFormContainer } from './AuthFormContainer';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';
import { ErrorMessage } from '../../../../core/components/atoms/ErrorMessage';
import { Alert } from '../../../../core/components/atoms/Alert';
import { AuthType } from '../../../../core/lib/utils/authType';

interface Props {
  onEmailChange: (value: string) => void;
  onPasswordChange: (value: string) => void;
  onConfirmPasswordChange: (value: string) => void;
  onFormSubmit: () => void;
  email: string;
  password: string;
  confirmPassword: string;
  authType: string;
  formError: string;
  authError: string;
}

export const AuthForm: React.FC<Props> = ({
  onEmailChange,
  onPasswordChange,
  onConfirmPasswordChange,
  onFormSubmit,
  email,
  password,
  confirmPassword,
  authType,
  formError,
  authError,
}) => (
  <AuthFormContainer>
    <Form>
      <Form.Group>
        <Form.Label>Email</Form.Label>
        <Form.Control
          type="email"
          name="email"
          placeholder="Enter Email"
          onChange={({ target }) => onEmailChange(target.value)}
          value={email}
        />
      </Form.Group>
      <Form.Group>
        <Form.Label>Password</Form.Label>
        <Form.Control
          type="password"
          name="password"
          placeholder="Enter Password"
          onChange={({ target }) => onPasswordChange(target.value)}
          value={password}
        />
      </Form.Group>
      {authType === AuthType[0] ? (
        <Form.Group>
          <Form.Label>Confirm Password</Form.Label>
          <Form.Control
            type="password"
            name="confirmPassword"
            placeholder="Enter Confirm Password"
            onChange={({ target }) => onConfirmPasswordChange(target.value)}
            value={confirmPassword}
          />
        </Form.Group>
      ) : null}
      {formError.length ? (
        <FlexBox>
          <ErrorMessage>{formError}</ErrorMessage>
        </FlexBox>
      ) : null}

      <FlexBox margin="10% 0 0 0">
        <Button
          onClick={onFormSubmit}
          disabled={formError.length > 0 || !email || !password}
          variant="primary"
        >
          Submit
        </Button>
      </FlexBox>
      {authError ? (
        <FlexBox margin="15px 0 5px 0">
          <Alert variant="danger">
            {authType} Error: {authError}
          </Alert>
        </FlexBox>
      ) : null}
    </Form>
  </AuthFormContainer>
);
