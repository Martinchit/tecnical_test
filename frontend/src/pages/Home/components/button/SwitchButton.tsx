import React from 'react';
import { ButtonContainer } from './ButtonContainer';
import { FlexBox } from '../../../../core/components/atoms/FlexBox';

interface Props {
  authType: string;
  onClick: () => void;
}

export const SwitchButton: React.FC<Props> = ({ authType, onClick }) => (
  <FlexBox margin="10px 0">
    <ButtonContainer id="change_auth_type_button" onClick={onClick} variant="light">
      {authType === 'Log In' ? 'Change to Sign Up' : 'Change to Log In'}
    </ButtonContainer>
  </FlexBox>
);
