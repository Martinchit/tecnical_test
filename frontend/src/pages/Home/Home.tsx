import React from 'react';
import { AuthForm } from './components/form/AuthForm';
import { SwitchButton } from './components/button/SwitchButton';
import { FlexBox } from '../../core/components/atoms/FlexBox';
import { PageTitle } from '../../core/components/molecules/PageTitle';
import { AuthType } from '../../core/lib/utils/authType';
import { AuthMode } from '../../types/types';

interface HomeState {
  email: string;
  password: string;
  confirmPassword: string;
  authType: AuthMode;
  formError: string;
}

interface HomeProps {
  logInAction: (email, password) => void;
  signUpAction: (email, password) => void;
  authError: string;
}

export class Home extends React.Component<HomeProps, HomeState> {
  constructor(props) {
    super(props);
    this.state = {
      email: '',
      password: '',
      confirmPassword: '',
      authType: AuthType[0],
      formError: '',
    };
  }

  onAuthTypeChangeHandler = () => {
    const { authType } = this.state;
    const newType = AuthType.indexOf(authType) ? AuthType[0] : AuthType[1];
    this.setState({
      authType: newType,
      email: '',
      password: '',
      confirmPassword: '',
      formError: '',
    });
  };

  onConfirmPasswordChange = (confirmPassword) => {
    this.setState({ confirmPassword }, () => {
      this.checkFormValid();
    });
  };

  onPasswordChange = (password) => {
    this.setState({ password }, () => {
      this.checkFormValid();
    });
  };

  onEmailChange = (email) => {
    this.setState({ email }, () => {
      this.checkFormValid();
    });
  };

  checkFormValid = () => {
    const { email, password, confirmPassword, authType } = this.state;
    const emailRegex = /^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/;
    const passwordRegex = /[\d|\w]{8,}/;
    if (!emailRegex.test(email)) {
      return this.setState({ formError: 'Email has to be in email pattern' });
    }
    if (!passwordRegex.test(password)) {
      return this.setState({
        formError:
          'Password has to have at least 8 characters and contain only letters or numbers',
      });
    }

    if (AuthType.indexOf(authType) === 0 && password !== confirmPassword) {
      return this.setState({
        formError: 'Confirm Password has to be same as Password',
      });
    }
    this.setState({ formError: '' });
  };

  onFormSubmit = () => {
    const { authType, email, password } = this.state;
    const { signUpAction, logInAction } = this.props;
    if (AuthType.indexOf(authType)) {
      logInAction(email, password);
    } else {
      signUpAction(email, password);
    }
  };

  render() {
    const {
      email,
      password,
      confirmPassword,
      authType,
      formError,
    } = this.state;
    const { authError } = this.props;
    return (
      <FlexBox>
        <PageTitle title={authType} />
        <AuthForm
          email={email}
          password={password}
          confirmPassword={confirmPassword}
          onConfirmPasswordChange={this.onConfirmPasswordChange}
          onPasswordChange={this.onPasswordChange}
          onEmailChange={this.onEmailChange}
          onFormSubmit={this.onFormSubmit}
          authType={authType}
          formError={formError}
          authError={authError}
        />
        <SwitchButton
          authType={authType}
          onClick={this.onAuthTypeChangeHandler}
        />
      </FlexBox>
    );
  }
}
