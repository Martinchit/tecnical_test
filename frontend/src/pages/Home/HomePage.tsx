import { Dispatch } from 'redux';
import { connect } from 'react-redux';
import { ActionsType, RootStateType, actions } from '../../redux/store';
import { Home } from './Home';

interface OwnProps {}

const mapStateToProps = (state: RootStateType) => ({
  authError: state.auth.error,
});

const mapDispatchToProps = (
  dispatch: Dispatch<ActionsType>,
  props: OwnProps
) => ({
  logInAction: (email, password) => dispatch(actions.logIn(email, password)),
  signUpAction: (email, password) => dispatch(actions.signUp(email, password)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Home);
