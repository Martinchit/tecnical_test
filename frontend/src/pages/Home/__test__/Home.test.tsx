import React from 'react';
import { shallow } from 'enzyme';

import { Home } from '../Home';
import { AuthForm } from '../components/form/AuthForm';
import { SwitchButton } from '../components/button/SwitchButton';

describe('Home', () => {
  const email = 'test@test.com';
  const password = '12345678';
  const confirmPassword = '12345678';
  const waitForAsync = () => new Promise((resolve) => setImmediate(resolve));
  const props = {
    logInAction: jest.fn(),
    signUpAction: jest.fn(),
    authError: '',
  };
  it('renders correctly', () => {
    const wrapper = shallow(<Home {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('handles onEmailChange', async () => {
    const email = 'test';
    const wrapper = shallow(<Home {...props} />);
    wrapper.find(AuthForm).first().props().onEmailChange(email);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.email).toBe(email);
    expect(stockListWrapper.state.formError).toBe(
      'Email has to be in email pattern'
    );
  });

  it('handles onEmailChange with correct pattern', async () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.find(AuthForm).first().props().onEmailChange(email);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.email).toBe(email);
    expect(stockListWrapper.state.formError).toBe(
      'Password has to have at least 8 characters and contain only letters or numbers'
    );
  });

  it('handles onPasswordChange', async () => {
    const password = '1234';
    const wrapper = shallow(<Home {...props} />);
    wrapper.find(AuthForm).first().props().onPasswordChange(password);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.password).toBe(password);
    expect(stockListWrapper.state.formError).toBe(
      'Email has to be in email pattern'
    );
  });

  it('handles onPasswordChange with correct pattern', async () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ email });
    wrapper.find(AuthForm).first().props().onPasswordChange(password);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.password).toBe(password);
    expect(stockListWrapper.state.formError).toBe(
      'Confirm Password has to be same as Password'
    );
  });

  it('handles onPasswordChange with correct pattern in Log In mode', async () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ email, authType: 'Log In' });
    wrapper.find(AuthForm).first().props().onPasswordChange(password);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.password).toBe(password);
    expect(stockListWrapper.state.formError).toBe('');
  });

  it('handles onPasswordChange', async () => {
    const confirmPassword = '1234';
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ email, password });
    wrapper
      .find(AuthForm)
      .first()
      .props()
      .onConfirmPasswordChange(confirmPassword);
    const stockListWrapper = wrapper.instance() as Home;
    await waitForAsync();
    expect(stockListWrapper.state.password).toBe(password);
    expect(stockListWrapper.state.formError).toBe(
      'Confirm Password has to be same as Password'
    );
  });

  it('handles onFormSubmit for signUp', async () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ email, password, confirmPassword });
    wrapper.find(AuthForm).first().props().onFormSubmit();
    expect(wrapper.find(AuthForm).first().props().formError).not.toBeTruthy();
    expect(props.signUpAction).toHaveBeenCalled();
    expect(props.signUpAction).toHaveBeenCalledWith(email, password);
  });

  it('handles onFormSubmit for logIn', async () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ email, password, confirmPassword, authType: 'Log In' });
    wrapper.find(AuthForm).first().props().onFormSubmit();
    expect(wrapper.find(AuthForm).first().props().formError).not.toBeTruthy();
    expect(props.logInAction).toHaveBeenCalled();
    expect(props.logInAction).toHaveBeenCalledWith(email, password);
  });

  it('handles onAuthTypeChangeHandler', () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.find(SwitchButton).first().props().onClick();
    const stockListWrapper = wrapper.instance() as Home;
    expect(stockListWrapper.state.authType).toBe('Log In');
  });

  it('handles onAuthTypeChangeHandler', () => {
    const wrapper = shallow(<Home {...props} />);
    wrapper.setState({ authType: 'Log In' });
    wrapper.find(SwitchButton).first().props().onClick();
    const stockListWrapper = wrapper.instance() as Home;
    expect(stockListWrapper.state.authType).toBe('Sign Up');
  });
});
