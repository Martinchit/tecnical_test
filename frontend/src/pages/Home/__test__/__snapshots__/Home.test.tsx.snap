// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
<styled.div>
  <Component
    title="Sign Up"
  />
  <Component
    authError=""
    authType="Sign Up"
    confirmPassword=""
    email=""
    formError=""
    onConfirmPasswordChange={[Function]}
    onEmailChange={[Function]}
    onFormSubmit={[Function]}
    onPasswordChange={[Function]}
    password=""
  />
  <Component
    authType="Sign Up"
    onClick={[Function]}
  />
</styled.div>
`;
